controllers/dspipeline_controller.go

```diff
@@ -501,6 +508,46 @@ func (r *DSPAReconciler) PublishMetrics(dspa *dspav1alpha1.DataSciencePipelinesA
 	}
 }
 
+func (r *DSPAReconciler) GetComponents(ctx context.Context, dspa *dspav1alpha1.DataSciencePipelinesApplication) dspav1alpha1.ComponentStatus {
+	log := r.Log.WithValues("namespace", dspa.Namespace).WithValues("dspa_name", dspa.Name)
+	log.Info("Updating components endpoints")
+
+	mlmdProxyUrl, err := util.GetServiceHostname(ctx, "ds-pipeline-md-"+dspa.Name, dspa.Namespace, r.Client)
```

Comment:
imho we should not even set a value if the service/route is not available. Do you think the code as is might end up with setting a value even though the service/route is not available?

Reply:
Actually, the code sets an empty string.

Reply:
yeah -- if the service isn't up, `util.GetServiceHostname` will return an empty string and no error. Which means below on line 526, `Url:` is going to be set to that empty string.

> imho we should not even set a value if the service/route is not available

agree -- so you'll need to do things piece by piece below instead of all in one shot.

For example (I didn't test this):
```
func (r *DSPAReconciler) GetComponents(ctx context.Context, dspa *dspav1alpha1.DataSciencePipelinesApplication) dspav1alpha1.ComponentStatus {
	log := r.Log.WithValues("namespace", dspa.Namespace).WithValues("dspa_name", dspa.Name)
	log.Info("Updating components endpoints")

	mlmdProxyUrl, err := util.GetServiceHostname(ctx, "ds-pipeline-md-"+dspa.Name, dspa.Namespace, r.Client)
	if err != nil {
		log.Error(err, "Error retrieving MLMD Proxy Service endpoint")
	}

	mlmdProxyExternalUrl, err := util.GetRouteHostname(ctx, "ds-pipeline-md-"+dspa.Name, dspa.Namespace, r.Client)
	if err != nil {
		log.Error(err, "Error retrieving MLMD Proxy Route endpoint")
	}

	apiServerUrl, err := util.GetServiceHostname(ctx, "ds-pipeline-"+dspa.Name, dspa.Namespace, r.Client)
	if err != nil {
		log.Error(err, "Error retrieving API Server Service endpoint")
	}

	apiServerExternalUrl, err := util.GetRouteHostname(ctx, "ds-pipeline-"+dspa.Name, dspa.Namespace, r.Client)
	if err != nil {
		log.Error(err, "Error retrieving API Server Route endpoint")
	}

	mlmdProxyComponent := &dspav1alpha1.ComponentDetailStatus{}
	if mlmdProxyUrl != "" {
		mlmdProxyComponent.Url = mlmdProxyUrl
	}
	if mlmdProxyExternalUrl != "" {
		mlmdProxyComponent.ExternalUrl = mlmdProxyExternalUrl
	}

	apiServerComponent := &dspav1alpha1.ComponentDetailStatus{}
	if apiServerUrl != "" {
		apiServerComponent.Url = apiServerUrl
	}
	if apiServerExternalUrl != "" {
		apiServerComponent.ExternalUrl = apiServerExternalUrl
	}
	
	status := dspav1alpha1.ComponentStatus{}
	if mlmdProxyComponent.Url != "" && mlmdProxyComponent.ExternalUrl != "" {
		status.MLMDProxy = *mlmdProxyComponent
	}
	if apiServerComponent.Url != "" && apiServerComponent.ExternalUrl != "" {
		status.APIServer = *apiServerComponent
	}
	return status	
}
```